File: 0160-intersection-of-two-linked-lists-相交链表/0160.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None

        B2A = False
        A2B = False
        pA = headA
        pB = headB

        while pA != pB:
            pA = pA.next
            pB = pB.next
            if not pA:
                if not A2B:
                    A2B = True
                    pA =  headB
                else:
                    return None
            if not pB:
                if not B2A:
                    B2A = True
                    pB = headA
                else:
                    return None
        
        return pA

File: 0160-intersection-of-two-linked-lists-相交链表/0160.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) return NULL;
        ListNode *p1 = headA;
        ListNode *p2 = headB;
        bool b1 = true;
        bool b2 = true;
        while (true) {
            if (!p1) {
                if (b1) {
                    p1 = headB;
                    b1 = false;
                }
                else return NULL;
            }
            if (!p2) {
                if (b2) {
                    p2 = headA;
                    b2 = false;
                }
                else return NULL;
            }
            if (p1 == p2) return p1;
            p1 = p1->next;
            p2 = p2->next;
        }
        return NULL;
    }
};